filebeat/input/journald/pkg/journalctl/reader.go

```diff
@@ -239,49 +316,91 @@ func (r *Reader) Next(cancel input.Canceler) (JournalEntry, error) {
 			} else {
 				r.backoff.Wait()
 			}
+		}
+
+		// If journalctl finished reading the messages from previous boots
+		// and exited successfully
+		if finished {
+			restart = true
+			extraArgs = append(extraArgs, "--follow")
+			if r.cursor != "" {
+				// If there is a cursor, only use the cursor and the follow argument
+				extraArgs = append(extraArgs, "--after-cursor", r.cursor)
+			} else {
+				// If there is no cursor, it means the first successfully run
+				// did not return any event, so we have to restart with the
+				// --follow and all the initial args.
+
+				extraArgs = append(extraArgs, r.firstRunArgs...)
+			}
+
+			r.state = followingState
+			r.logger.Info("finished reading journal entries from all boots, restarting journalctl with follow flag")
+		}
 
-			jctl, err := r.jctlFactory(r.canceler, r.logger.Named("journalctl-runner"), "journalctl", args...)
-			if err != nil {
+		// Restart journalctl if needed
+		if restart {
+			if err := r.newJctl(extraArgs...); err != nil {
 				// If we cannot restart journalct, there is nothing we can do.
 				return JournalEntry{}, fmt.Errorf("cannot restart journalctl: %w", err)
 			}
-			r.jctl = jctl
 
 			// Return an empty message and wait for the input to call us again
 			return JournalEntry{}, ErrRestarting
 		}
 	}
 
+	return r.handleMessage(msg)
+}
+
+func (r *Reader) handleMessage(msg []byte) (JournalEntry, error) {
 	fields := map[string]any{}
-	if err := json.Unmarshal(d, &fields); err != nil {
-		r.logger.Error("journal event cannot be parsed as map[string]any, look at the events log file for the raw journal event")
+	if err := json.Unmarshal(msg, &fields); err != nil {
+		r.logger.Error("journal event cannot be parsed as map[string]any, " +
+			"look at the events log file for the raw journal event")
+
 		// Log raw data to events log file
-		msg := fmt.Sprintf("data cannot be parsed as map[string]any JSON: '%s'", string(d))
-		r.logger.Errorw(msg, logp.TypeKey, logp.EventType, "error.message", err.Error())
+		msg := fmt.Sprintf("data cannot be parsed as map[string]any. Data: '%s'",
+			string(msg))
+		r.logger.Errorw(
+			msg,
+			"error.message", err.Error(),
+			logp.TypeKey, logp.EventType)
+
 		return JournalEntry{}, fmt.Errorf("cannot decode Journald JSON: %w", err)
 	}
 
 	ts, isString := fields["__REALTIME_TIMESTAMP"].(string)
 	if !isString {
-		return JournalEntry{}, fmt.Errorf("'__REALTIME_TIMESTAMP': '%[1]v', type %[1]T is not a string", fields["__REALTIME_TIMESTAMP"])
+		return JournalEntry{},
+			fmt.Errorf("'__REALTIME_TIMESTAMP': '%[1]v', type %[1]T is not a string",
+				fields["__REALTIME_TIMESTAMP"])
 	}
 	unixTS, err := strconv.ParseUint(ts, 10, 64)
 	if err != nil {
-		return JournalEntry{}, fmt.Errorf("could not convert '__REALTIME_TIMESTAMP' to uint64: %w", err)
+		return JournalEntry{},
+			fmt.Errorf("could not convert '__REALTIME_TIMESTAMP' to uint64: %w",
+				err)
 	}
 
 	monotomicTs, isString := fields["__MONOTONIC_TIMESTAMP"].(string)
```

Comment:
NIT: These field names could be constants, if they are standdard and references in more than one place.

Reply:
Makes sense, thanks!